ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java

```diff
@@ -121,4 +110,111 @@ public void start() throws ReplicationException {
       logger.error("IO error while starting replication", e);
     }
   }
+
+  /**
+   * Add given replica into replication manager.
+   * @param replicaId the replica to add
+   * @return {@code true} if addition succeeded, {@code false} failed to add replica because it already exists.
+   */
+  public boolean addReplica(ReplicaId replicaId) {
+    boolean succeed = false;
+    rwLock.readLock().lock();
+    try {
+      if (partitionToPartitionInfo.containsKey(replicaId.getPartitionId())) {
+        logger.error("{} already exists in replication manager, rejecting adding replica request.",
+            replicaId.getPartitionId());
+      } else {
+        List<? extends ReplicaId> peerReplicas = replicaId.getPeerReplicaIds();
+        List<RemoteReplicaInfo> remoteReplicaInfos = new ArrayList<>();
+        if (peerReplicas != null) {
+          remoteReplicaInfos = createRemoteReplicaInfos(peerReplicas, replicaId);
+          updatePartitionInfoMaps(remoteReplicaInfos, replicaId);
+        }
+        logger.info("Assigning thread for {}", replicaId.getPartitionId());
+        addRemoteReplicaInfoToReplicaThread(remoteReplicaInfos, true);
+        // No need to update persistor to explicitly persist tokens for new replica because background persistor will
+        // periodically persistor all tokens including new added replica's
+        // TODO ensure there is a test that checks persistor has new replica info.
+        logger.info("{} is successfully added into replication manager", replicaId.getPartitionId());
+        succeed = true;
+      }
+    } finally {
+      rwLock.readLock().unlock();
+    }
+    return succeed;
+  }
+
+  /**
+   * Remove replica from replication manager
+   * @param replicaId the replica to remove
+   * @return {@code true} if replica is successfully removed. {@code false} otherwise
+   */
+  public boolean removeReplica(ReplicaId replicaId) {
+    boolean succeed = false;
+    rwLock.readLock().lock();
```

Comment:
why do all of these methods that mutate the data structures take the read lock? shouldn't they take the write lock since they mutate the data structure?

Reply:
I don't see any usages of the write lock in this class. Typically a readwritelock is used to allow for concurrent readers during normal scenarios (by using the read lock), but lock the entire datastructure (ensure that all readers finish there reads and make future readers wait by using the write lock) when editing.

Reply:
right, will change this to simple `ReentrantLock`.

Reply:
With just a simple mutex, wouldn't you have to synchronize all reads/writes to these data structures? Would it be better to take a read lock when reading from the shared data structures and taking a write lock when editing them? We can discuss in person.